### PR TITLE
Style Guide Alignment 

### PR DESCRIPTION
**Pull request by chrisvfritz**
I believe this is [everything that's needed](https://github.com/vuejs/eslint-plugin-vue/issues/77#issuecomment-333954255) to get the plugin to match the style guide. As I worked, I also fixed any minor bugs or inconsistencies I noticed.

### CODE REVIEW

**Comment by chrisvfritz** [comment]
I'd rather focus on the rules we have for now. I'm mostly holding off on adding new rules to the style guide too, while we refine what we currently have. And we'll be adding references later to note what has a corresponding style guide rule here - and on the style guide, what's enforceable by ESLint.

I figure it's better to have solid set of rules available than nothing, as we wait for every possible rule to be implemented. 🙂

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,4 @@
 /tests/integrations/*/node_modules
 /node_modules
 /test.*
+yarn.lock
```
Comment (michalsnik) [inline .gitignore]: I wouldn't ignore yarn.lock, we might want to add it to repository

File: README.md [hunk 2]
```diff
@@ -75,85 +75,107 @@ Vue.component('AsyncComponent', (resolve, reject) => {
 ## :gear: Configs
 
 This plugin provides two predefined configs:
-- `plugin:vue/base` - contains necessary settings for this plugin to work properly
-- `plugin:vue/recommended` - extends base config with recommended rules (the ones with check mark :white_check_mark: in the table below)
+- `plugin:vue/base` - Settings and rules to enable correct ESLint parsing
+- `plugin:vue/essential` - Above, plus rules to prevent errors or unintended behavior
```
Comment (michalsnik) [inline README.md]: Those configs seems very nice, but I'm not sure we should group them like this in readme. Current groups are common in eslint ecosystem (take a look at original rules https://eslint.org/docs/rules/). Plus later we will probably have some rules that we don't want to enable in any config and give the user full freedom in deciding whether he needs it or not. With the proposed granulation we would also have to add another category - which may lead to unnecessary churn. Also I'm not sure whether maintaining 4 configs is a way to go. I'd rather keep it  simple stupid and provide only `base` and `recommended` config. What do you think @chrisvfritz @mysticatea ?

File: docs/rules/name-property-casing.md [hunk 3]
```diff
@@ -22,18 +22,10 @@ export default {
 }
 ```
 
-:+1: Examples of **correct** code for `camelCase`:
```
Comment (michalsnik) [inline docs/rules/name-property-casing.md]: Why did you remove this? Recommending something is one thing, but we should still have the ability to configure rule as we see fit.

File: lib/config/all.js [hunk 4]
```diff
@@ -0,0 +1,4 @@
+module.exports = {
```
Comment (michalsnik) [inline lib/config/all.js]: what do we need this config for?

**Comment by michalsnik** [comment]
Great work @chrisvfritz! 🚀  I have concerns though, regarding proposed categories and number of configs ☝️

File: .gitignore (hunk 1 above)
Comment (chrisvfritz) [reply .gitignore]: I thought it'd be safer to add it to `.gitignore` until an explicit decision to include it had been made, at which time it could be easily removed. Do you disagree?

File: README.md (hunk 2 above)
Comment (chrisvfritz) [reply README.md]: Those configs seems very nice, but I'm not sure we should group them like this in readme. Current groups are common in eslint ecosystem (take a look at original rules eslint.org/docs/rules). Plus later we will probably have some rules that we don't want to enable in any config and give the user full freedom in deciding whether he needs it or not. With the proposed granulation we would also have to add another category - which may lead to unnecessary churn. Also I'm not sure whether maintaining 4 configs is a way to go. I'd rather keep it simple stupid and provide only base and recommended config. What do you think @chrisvfritz @mysticatea ?

===

> Current groups are common in eslint ecosystem (take a look at original rules eslint.org/docs/rules).

I agree it makes sense to stick with ESLint conventions wherever there's not a compelling reason to stray, but I did see some compelling reasons here.

First, I'm thinking that most people will probably discover the ESLint plugin through the style guide, so I saw a strong benefit in not forcing users to context switch after arriving here. If they want to enforce priority A rules, they'll know within seconds exactly what to do. There will be no wondering, "Wait, is Possible Errors what I want to enforce Priority A rules?"

For users who have never seen the style guide, but are familiar with ESLint conventions, I don't think the new name will cause any confusion - but perhaps there's something I'm not foreseeing. What do you think?

Second, I think there's value in the increased granularity. "Stylistic Issues" is too broad in my opinion, which is why I split stylistic rules between Priority B and C. For Priority B, there's one or more possible conventions that offer clear benefits over the alternatives. For Priority C, there are really many good options and the primary value is just in picking something and sticking to it.

Finally, I moved `jsx-uses-vars` out of "Variables" because I just don't find that a useful category, despite ESLint's recommendation. Their rules in that category include a mix of possible errors (e.g. `no-use-before-define`) and stylistic issues (e.g. `init-declarations`) and this is the only language feature (apart from ES6) that receives special treatment. I also feel that `jsx-uses-vars` only _incidentally_ has to do with variables - it's really a parsing extension that allows ESLint to work correctly, which is why I added it to `base`.

> Plus later we will probably have some rules that we don't want to enable in any config and give the user full freedom in deciding whether he needs it or not.

I'm not sure I understand this one. Don't users have the same amount of freedom either way, since they can override whatever they want? In terms of ease of use, I can't think of a case where more nested rule sets would worsen the user experience or make it more difficult for them to do something. Am I missing something though?

> With the proposed granulation we would also have to add another category - which may lead to unnecessary churn.

I might need more context to understand this one, because adding a new category wouldn't be a lot more code nor increase its complexity, if I'm understanding correctly. Can you provide a specific example of what you're thinking might happen?

> Also I'm not sure whether maintaining 4 configs is a way to go. I'd rather keep it simple stupid and provide only base and recommended config.

This is another one where I might need more context. The configs are generated with a script, so I don't believes there's any extra maintenance needed.

File: docs/rules/name-property-casing.md (hunk 3 above)
Comment (chrisvfritz) [reply docs/rules/name-property-casing.md]: Not necessarily. For example, the `new-cap` and `camelcase` rules don't include the ability to enforce `snake_case`. Some might argue they should be able to configure variable and constructor casing however they like, but if there's no use case, it's not included as an option.

Similarly, there are benefits for PascalCase and kebab-case in the `name` property, but I don't believe any scenario exists where you'd want to use camelCase. It would simply break convention both in JavaScript (constructors are PascalCase) and HTML (custom elements are kebab-case).

File: lib/config/all.js (hunk 4 above)
Comment (chrisvfritz) [reply lib/config/all.js]: Oops. 😅  That was something I experimented with, but decided there wasn't a use case for simply including all rules. I've deleted it now. Good catch!

**Comment by chrisvfritz** [comment]
@michalsnik Thanks for the feedback! 🙂  I responded to your concerns to provide more information. Sorry for not including that context in the PR description. 😅

**Comment by chrisvfritz** [comment]
@michalsnik Just pinging on this. 🙂  I tried to address your concerns above and would love to hear your thoughts.

File: docs/rules/attribute-hyphenation.md [hunk 5]
```diff
@@ -15,41 +15,25 @@ Default casing is set to `always`
 :+1: Examples of **correct** code`:
 
 ```html
-<template>
-  <foo my-prop="prop">
-    <a onClick="return false"></a>
-  </foo>
-</template>
+<MyComponent my-prop="prop"/>
```
[the remark this answers is not in the document]
Comment (chrisvfritz) [reply docs/rules/attribute-hyphenation.md]: @Aladdin-ADD See [this rule](https://vuejs.org/v2/style-guide/#Component-name-casing-in-templates-strongly-recommended) in the style guide.

File: docs/rules/name-property-casing.md (hunk 3 above)
Comment (michalsnik) [reply docs/rules/name-property-casing.md]: Okay, if it's aligned with the rule now then cool :)

File: README.md (hunk 2 above)
Comment (michalsnik) [reply README.md]: Okay, I get your point better now @chrisvfritz Thanks for adding more context :)

And regarding last three paragraphs that you quoted:
I was thinking about the case where we want to introduce a rule, but we don't want to enable it in any config by default. Let's say we'd want to give people possibility to use the given rule, but it's not aligned with the official styleguide for some reason.
Maybe it's an edge case, but we had a similar scenario in eslint-plugin-ember.

Adding another group is not a big deal though, so I think that we can move forward with the current state, and iterate over it later on 👍

**Comment by michalsnik** [comment]
Yeah, you changed my mind @chrisvfritz :) Since now we have an official style guide and this is also an official plugin for Vue, it makes sense to be more opinionated and to have the rules organised as proposed. However before merging I would love to hear what @mysticatea thinks about it as well :)

File: README.md (hunk 2 above)
Comment (chrisvfritz) [reply README.md]: @michalsnik I agree it'd be good to introduce a category like that when the need arises - and I can already think of use cases for it. For example, I like to disallow filters in many projects, because they add one more thing that developers have to learn without enabling them to do anything new. That kind of thing wouldn't fit into any (existing) categories of the style guide though.

**Comment by chrisvfritz** [comment]
@michalsnik Agreed. 🙂 👍  Just added a request for his review.

**Comment by michalsnik** [comment]
Thanks for the last fix @chrisvfritz, you were right :) Please rebase/merge master now, and @mysticatea can you please take a look at this PR as well?

**Comment by mysticatea** [comment]
I apology that I'm inactive during a long time.

Wow, amazing work!
Looks almost good to me.

One point, I have a concern that this PR removed `recommended: true/false` property of all rules.
Because [our semver policy](https://github.com/vuejs/eslint-plugin-vue#anchor-semantic-versioning-policy) is following [ESLint semver policy](https://github.com/eslint/eslint#semantic-versioning-policy), so it's `semver-major` matter if we add a new rule into configs like `vue/recommended`. I think we need the way to add rules without config changes in order to avoid major bump rushes.

**Comment by chrisvfritz** [comment]
Yes, thanks for the ping! I agree with you @mysticatea and will update this weekend. And while I think that semver policy makes sense, I also think it might not be too bad to publish major versions relatively often (e.g. as often as monthly?), especially in the beginning as there may be many new rules to add. Do you think that would be OK?

**Comment by mysticatea** [comment]
Thank you, @chrisvfritz.

I agree with you. I think that one or two of new `Essential` rules is valuable to make a major release. ESLint itself has a policy they don't make two major releases within six months, but off course, plugins can make major releases more frequency. Monthly or bimonthly looks good to me.

**Comment by michalsnik** [comment]
Hey @chrisvfritz would you be so kind and resolve these conflicts? :) I think we should start merging it all together now

**Comment by chrisvfritz** [comment]
@michalsnik Conflicts resolved! 😄  

@mysticatea Rules no longer require a category and if they don't, they'll be added to an "Uncategorized" section after the main rules in docs, but before the "Deprecated" rules.

I also updated the default for html-indent rules from 4 spaces to 2 spaces, since that's what's used throughout all the Vue docs and just seems to be a more popular in the web community (e.g. GitHub's style guide, Google's style guide, and [in GitHub repos](https://ukupat.github.io/tabs-or-spaces/)).

**Comment by mysticatea** [comment]
I'm sorry for late: `Uncategorized` looks reasonable to me.
